libp2p/protocols/connectivity/dcutr/server.nim

```diff
@@ -0,0 +1,55 @@
+# Nim-LibP2P
+# Copyright (c) 2022 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import std/[options, sets, sequtils]
+
+import core
+import ../../protocol,
+       ../../../stream/connection,
+       ../../../switch
+
+import stew/[results, objects]
+import chronos, chronicles
+
+export chronicles
+
+type Dcutr* = ref object of LPProtocol
+
+logScope:
+  topics = "libp2p dcutr"
+
+proc new*(T: typedesc[Dcutr], switch: Switch): T =
+
+  proc handleStream(stream: Connection, proto: string) {.async, gcsafe.} =
+    try:
+      let connectMsg = DcutrMsg.decode(await stream.readLp(1024))
+      debug "Dcutr receiver received a Connect message.", connectMsg
+      var ourAddrs = switch.peerStore.getMostObservedProtosAndPorts() # likely empty when the peer is reachable
+      if ourAddrs.len == 0:
+        # this list should be the same as the peer's public addrs when it is reachable
+        ourAddrs =  switch.peerInfo.addrs.mapIt(switch.peerStore.guessDialableAddr(it))
```

Comment:
Can we get in this situation when we are not reachable? If so, how does it handle circuit relay addresses & co?

Reply:
Instinctively it seems like this should be `switch.peerInfo.listenAddrs` instead of addrs

Reply:
> Can we get in this situation when we are not reachable?
Yes, we'll get here when there aren't enough observations (currently at least 3) of the same IP/Port.

Reply:
> how does it handle circuit relay addresses & co?

I handled this in the HP PR, but should have done it here. Does it make sense?
```
      var dialableAddrs = connectMsg.addrs
      dialableAddrs.keepItIf(not it.contains(multiCodec("p2p-circuit")).valueOr(true))
      if dialableAddrs.len == 0:
        debug "Dcutr receiver has no dialable addresses to connect to."
        return
```

Reply:
Now I realized you were asking about sending relay addrs. You're right, it should be `listenAddrs`.